spec/acceptance/realtime/auth_spec.rb

```diff
@@ -1031,5 +1031,184 @@ def disconnect_transport(connection)
+        let(:client_options) { { token: token.body, environment: environment, protocol: protocol } }
+
+        context 'when credentials are valid' do
+          let(:token) { Faraday.get "#{auth_url}?keyName=#{key_name}&keySecret=#{key_secret}" }
+
+          it 'pulls stats successfully' do
+            client.stats do |stats|
+              expect(stats).to_not be nil
+              stop_reactor
+            end
+          end
+        end
+
+        context 'when credentials are invalid' do
+          let(:token) { Faraday.get "#{auth_url}?keyName=#{key_name}&keySecret=invalid" }
+
+          it 'fails with an invalid signature error' do
+            client.connection.once(:disconnected) do |state_change|
+              expect(state_change.reason.message.match(/invalid signature/i)).to_not be_nil
+              expect(state_change.reason.code).to eql(40144)
+              stop_reactor
+            end
+            client.connect
+          end
+        end
+      end
+
+      context 'when JWT token expires' do
+        before do
+          stub_const 'Ably::Models::TokenDetails::TOKEN_EXPIRY_BUFFER', 0 # allow token to be used even if about to expire
+          stub_const 'Ably::Auth::TOKEN_DEFAULTS', Ably::Auth::TOKEN_DEFAULTS.merge(renew_token_buffer: 0) # Ensure tokens issued expire immediately after issue
+        end
+        let(:token_callback) do
+          lambda do |token_params|
+            # Ably in all environments other than production will send AUTH 5 seconds before expiry, so
+            # we generate a JWT that expires in 5s so that the window for Realtime to send has passed
+            tokenResponse = Faraday.get "#{auth_url}?keyName=#{key_name}&keySecret=#{key_secret}&expiresIn=5"
+            tokenResponse.body
+          end
+        end
+        let(:client_options) { default_options.merge(use_token_auth: true, auth_callback: token_callback) }
+
+        it 'a new token is requested via auth_callback and the client gets reconnected' do
+          client.connection.once(:connected) do
+            original_token = auth.current_token_details
+            original_conn_id = client.connection.id
+
+            client.connection.once(:disconnected) do |state_change|
+              expect(state_change.reason.code).to eql(40142)
+
+              client.connection.once(:connected) do
+                expect(original_token).to_not eql(auth.current_token_details)
+                expect(original_conn_id).to eql(client.connection.id)
+                stop_reactor
+              end
+            end
+          end
+        end
+      end
+
+      context 'when the JWT token request includes a client_id' do
+        let(:client_id) { random_str }
+        let(:auth_callback) do
+          lambda do |token_params|
+            Ably::Rest::Client.new(default_options).auth.request_token({ client_id: client_id }, { auth_url: auth_url, auth_params: auth_params }).token
```

Comment:
This feels wrong to me, but I stand to be corrected about what is happening here.  Given the `auth_url` is getting a JWT token, and that JWT token has a `client_id` embedded, then why we are bothering to a) pass in a client_id into the request `TokenParams` given the `auth_url` in reality would be responsible for that, b) why even use the REST library here, we are testing the realtime library here i.e. isn't it clearer to show explicitly what is returned from `auth_callback` whereas right now it's just not clear if that is an Ably token, a JWT token, etc.